### PR TITLE
(api) add oat++ lib

### PR DESCRIPTION
**Pull request by ddnowicki**

### CODE REVIEW

File: Testing/Temporary/CTestCostData.txt
Comment (patkaczm) [inline Testing/Temporary/CTestCostData.txt:1]: this file should not be added

File: Testing/Temporary/LastTest.log
Comment (patkaczm) [inline Testing/Temporary/LastTest.log:1]: this file should not be added

**Comment by patkaczm** [comment]
Even though adding oat the way you did it works, I'd suggest to make it cleaner and only check if library is visible for cmake via `find package` function and link it/ add `target_include_directories` to make it visible from compiler level.

**Comment by ddnowicki** [comment]
I have problem, when i add `find_package(oat REQUIRED)` i get an error:
```
vboxuser@Ubuntu:~/Documents/paragon_back$ cmake --build build 
CMake Error at src/api/CMakeLists.txt:3 (find_package):
  By not providing "Findoat.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "oat", but
  CMake did not find one.

  Could not find a package configuration file provided by "oat" with any of
  the following names:

    oatConfig.cmake
    oat-config.cmake

  Add the installation prefix of "oat" to CMAKE_PREFIX_PATH or set "oat_DIR"
  to a directory containing one of the above files.  If "oat" provides a
  separate development package or SDK, be sure it has been installed.


-- Configuring incomplete, errors occurred!
See also "/home/vboxuser/Documents/paragon_back/build/CMakeFiles/CMakeOutput.log".
gmake: *** [Makefile:688: cmake_check_build_system] Error 1
```
I tried adding Findoatpp.cmake file, but this file can't find oat++.
I delete test because it was useless.

**Comment by patkaczm** [comment]
> I have problem, when i add `find_package(oat REQUIRED)` i get an error:
> 
> ```
> vboxuser@Ubuntu:~/Documents/paragon_back$ cmake --build build 
> CMake Error at src/api/CMakeLists.txt:3 (find_package):
>   By not providing "Findoat.cmake" in CMAKE_MODULE_PATH this project has
>   asked CMake to find a package configuration file provided by "oat", but
>   CMake did not find one.
> 
>   Could not find a package configuration file provided by "oat" with any of
>   the following names:
> 
>     oatConfig.cmake
>     oat-config.cmake
> 
>   Add the installation prefix of "oat" to CMAKE_PREFIX_PATH or set "oat_DIR"
>   to a directory containing one of the above files.  If "oat" provides a
>   separate development package or SDK, be sure it has been installed.
> 
> 
> -- Configuring incomplete, errors occurred!
> See also "/home/vboxuser/Documents/paragon_back/build/CMakeFiles/CMakeOutput.log".
> gmake: *** [Makefile:688: cmake_check_build_system] Error 1
> ```
> 
> I tried adding Findoatpp.cmake file, but this file can't find oat++. I delete test because it was useless.

Have you installed the oatpp in your system using `make install` command?

**Comment by ddnowicki** [comment]
My mistake, I'm only installing the Oat++ src file instead of the full project
So I install using git clone
```
git clone https://github.com/oatpp/oatpp.git
cd oatpp
mkdir build
cd build
cmake ..
make
sudo make install
```

and then implement `find package` in src/api/CMakeLists.txt, and while compiling i get err
```
[ 65%] Building CXX object src/api/CMakeFiles/api-lib.dir/libs/oatpp/src/oatpp/web/server/AsyncHttpConnectionHandler.cpp.o
[ 65%] Building CXX object src/api/CMakeFiles/api-lib.dir/libs/oatpp/src/oatpp/web/server/HttpConnectionHandler.cpp.o
[ 66%] Building CXX object src/api/CMakeFiles/api-lib.dir/libs/oatpp/src/oatpp/web/server/HttpProcessor.cpp.o
[ 66%] Building CXX object src/api/CMakeFiles/api-lib.dir/libs/oatpp/src/oatpp/web/server/HttpRouter.cpp.o
[ 67%] Building CXX object src/api/CMakeFiles/api-lib.dir/libs/oatpp/src/oatpp/web/server/api/ApiController.cpp.o
[ 67%] Building CXX object src/api/CMakeFiles/api-lib.dir/libs/oatpp/src/oatpp/web/server/api/Endpoint.cpp.o
[ 68%] Building CXX object src/api/CMakeFiles/api-lib.dir/libs/oatpp/src/oatpp/web/server/handler/AuthorizationHandler.cpp.o
[ 69%] Building CXX object src/api/CMakeFiles/api-lib.dir/libs/oatpp/src/oatpp/web/server/handler/ErrorHandler.cpp.o
[ 69%] Building CXX object src/api/CMakeFiles/api-lib.dir/libs/oatpp/src/oatpp/web/server/interceptor/AllowCorsGlobal.cpp.o
[ 70%] Building CXX object src/api/CMakeFiles/api-lib.dir/libs/oatpp/src/oatpp/web/url/mapping/Pattern.cpp.o
[ 70%] Building CXX object src/api/CMakeFiles/api-lib.dir/libs/oatpp/test/oatpp/AllTestsMain.cpp.o
/home/vboxuser/Documents/paragon_back/src/api/libs/oatpp/test/oatpp/AllTestsMain.cpp:2:10: fatal error: oatpp/web/ClientRetryTest.hpp: No such file or directory
    2 | #include "oatpp/web/ClientRetryTest.hpp"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
gmake[2]: *** [src/api/CMakeFiles/api-lib.dir/build.make:1896: src/api/CMakeFiles/api-lib.dir/libs/oatpp/test/oatpp/AllTestsMain.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:273: src/api/CMakeFiles/api-lib.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
root@Ubuntu:/home/vboxuser/Documents/paragon_back# 
```

**Comment by patkaczm** [comment]
Now, when install step is done, the library is available in our system and we won't build it from the source. Instead we'll only `find` it in cmake and let cmake add proper includes via `target_include_directories` and link via `target_link_library`.
Check what `oatppConfig.cmake` sets and use it in `api/CMakeLists.txt` and `test/api/CMakeLists.txt` if needed.
If you have further questions/problems please reach me out on discord.

**Comment by patkaczm** [comment]
BTW, now whole dir /api/libs is superfluous.

File: src/api/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+find_package(oatpp 1.3.0 REQUIRED)
+
+add_library(api-lib Api.cpp)
+
+target_include_directories(api-lib PUBLIC ${oatpp_INCLUDE_DIRS} oatpp:oatpp)
```
Comment (patkaczm) [inline src/api/CMakeLists.txt]: Is this needed here?

**Review by patkaczm — APPROVED** [review]

**Review by DrzewJak — APPROVED** [review]

**Pull request merged**